### ISSUE TITLE
Set up a DMZ or isolated network to securely host services rather than tunnelling them

### ISSUE THREAD

**Issue by MatthewCroughan** [DoESLiverpool/somebody-should#1074]
I've already set up a storage solution at https://nextcloud.matthewcroughan.co.uk that allows access to the DoES local SMB share in a secure manner, meaning people can access their files at DoES, store their work files in a way that is instantly accessible, without clogging up the machines internal storage space (dropbox) and or thrashing their disks. 

The problem with this is that I'm finding myself doing this for other things now too, such as an RDP server on @ajlennon's machine that can give people access to a high powered machine inside or outside of DoES. Since RDP runs via UDP I can easily tunnel this out to a server that I control and then do the same again, but this isn't scalable or secure.

I would really appreciate it if we could accelerate an effort for an external/secondary network where we can run services.

#947

**Comment by ajlennon** [comment]
Hi all - I've discussed with Matt and I think what he's doing is really useful work.

I'm aware that security of the space is an issue though and I do believe we should probably ensure everybody agrees on direction before more holes are poked through firewalls and so forth?

**Comment by johnmckerrell** [comment]
On #817 a VPN was mentioned, how much of what you want to do could be achieved with a VPN? (Probably not next cloud but can you do the other things you've been trying to do?)

**Comment by MatthewCroughan** [comment]
@johnmckerrell 
Is there a reason why Baltic can not just DMZ our router and then we can port forward from there?

_Originally posted by @skos-ninja in https://github.com/DoESLiverpool/somebody-should/issues/817#issuecomment-416668425_

It'd be much nicer if we could co-ordinate this properly from the start, it's far more effort for an inferior result to try and use a VPN. We need a properly isolated network. What I'm doing is already effectively a VPN to my server.

**Comment by johnmckerrell** [comment]
> Just a thought is we should probably not port forward things ... and instead we should vpn into the network to then access that stuff?

*Also posted by @skos-ninja in #817* (yes I know I edited that to benefit myself)

Baltic DMZ-ing our router has nothing to do with how we arrange our network, it just means that we don't have to ask them to open each individual port.

I'm not saying we shouldn't open any ports, but if you want to be able to Remote Desktop or SSH into a machine, you should do that over a VPN connection, if you want to make a service such as Next Cloud available publicly then that wouldn't go on a VPN. I'm asking you (us) to think about what it is we want to make available.

At no point will it be possible for you to simply run a new service on your computer and have that available publicly instantly, certainly not via IPv4 (as we only have one IP address). There will always be port forwarding needed on our router.

**Comment by ajlennon** [comment]
+1 for VPN access to webcams and my computer and so forth

**Comment by MatthewCroughan** [comment]
@johnmckerrell Right, so currently I couldn't make use of a VPN to have things like Nextcloud open to the internet, I'd still have to poke holes through tunnels. Currently I have a kind-of VPN going on with my server externally, and that is all that allows me to do this. 

By providing a VPN into DoES, it does make it easier for me to interact with things in does, like my pi's, which I'm already doing via my server. Saves a lot of hassle. It definitely makes things better, and I'd be thankful for it, since I'd be able to get on with my stuff.

The only problem with that is we can't have anything be public facing via it, which means anybody wanting to use nextcloud outside of does would have to VPN into does, and we couldn't have a memorable domain associated with it.

**Comment by johnmckerrell** [comment]
Indeed, good thing I didn't say VPN was the only option! I don't believe we can just set up a network and say "DMZ!" and that means that you just plug things in and they're magically available, IPv4 won't allow it. You will always need to request that an external port is forwarded to your internal IP address (currently the DoES network would need tweaking and the Baltic Broadband one too but we might be able to get them to forward everything on to us so it's just our end that needed tweaking).

Whether "things that are on the internet" need to be on their own separate network I'm not sure, it probably wouldn't hurt but if you're talking about having the server on both networks then it doesn't really help, i.e. the reason you put it on its own network is "if it gets hacked then the other network isn't exposed" but if it's on both networks then.. it exposes both networks anyway.

**Comment by MatthewCroughan** [comment]
@johnmckerrell The only bridging that would occur would be sensible, for example SMB via Nextcloud. SMB in and of itself is a silly thing to expose to the internet, whereas Nextcloud is intended for public facing usage. I've not done this before, and I don't know how to make this happen, hence why I'm asking you and Jake to look into it.

One bad, old, vulnerable protocol exposed would mean that the network is compromised in some way, even with a VPN, which is why a separate network entirely makes more sense. 

A VPN would be less secure, since it would provide people access onto our main network, with no segregation between services, if it were used by a lot of people it could potentially invite more activity and potential risks I don't know of any other way to have public facing things in a secure way other than a DMZ, since I don't know much of anything about networking. It's the only term I know of that describes what I want, since again, I don't know enough about networking.

A VPN is a good short term solution for me personally, and I already have a small one via ssh tunnels, but I was going to set up wireguard/openvpn, but if we can get this sorted officially then I'd prefer to do that.

**Comment by johnmckerrell** [comment]
The new IP addresses are due to be setup on 7th May 6-7pm. During this time there will be some network downtime but hopefully it will be minimal. No promises when the new IP addresses will be available but we should be able to do it following this work.

**Comment by ajlennon** [comment]
What's the thinking on which network segment the new IP addresses will sit?

**Comment by johnmckerrell** [comment]
Current thinking is that we'd create new networks for each IP and these could be allocated to physical ports on the switch, so any device plugged into a certain ethernet port would be exposed on that IP address. We would require that these devices remained only on that network and that no bridging was performed (i.e. having a device both on an "external" network and on the DoES internal network).

**Comment by ajlennon** [comment]
Makes a lot of sense to me. I might be interested in having multiple devices serving different services on the public TCP ports if that can be accommodated...

**Comment by johnmckerrell** [comment]
For you to have control over that, the way to do it would be for you to have your own router that performs the NATing that would be needed here. The alternative would be for DoES to manage this and port forward specific services for specific IP addresses but you would not be able to make any changes to the setup yourself (I think so anyway, I should check if the unifi software would allow different users to have different levels of access, i.e. if I could give you access to our router but only to add settings to your own WAN that might be interesting).

**Comment by ajlennon** [comment]
To be fair I don't think I should be given access to more stuff I might break...

It would be easy enough for me to set up an RPi box as a NAT gateway for some other boxes if there's space somewhere...

**Comment by ajlennon** [comment]
Oh and if I pay more can I have a personalised IP address? 

Like 65.76.69.120 or something ?

**Comment by ajlennon** [comment]
13.79.69.83 would be cool too

**Comment by johnmckerrell** [comment]
Technically yes, I’m sure you can buy 2.2.2.2 if you pay someone enough money but I don’t know too much about that! We have a subnet assigned to us with 5 usable IP addresses so you’ll just get one of those (don’t have the range to hand right now).

**Comment by johnmckerrell** [comment]
The best way to set these IP addresses up is to place a switch between the Baltic router and our Ubiquiti gateway. Ideally it would be a managed switch with an IP address assigned to each port. Each person who has paid for an IP address would then get assigned a port on the switch and could plug what they need into it, whether a machine, or a router that could then do further NAT-ing. The simplest thing would be having the hosted machine within the cupboard but if we're careful we could potentially use the existing wiring to wire one of these IP addresses into the main room, although it would be much more preferable to have them in the cupboard as we know how people like to juggle ethernet cables around.

@ajlennon have a read of that and give me a shout about what you'd like to do.

**Comment by ajlennon** [comment]
All sounds good to me.

I am sure I brought in a decent Netgear managed switch some time ago. Any ideas where that went @amcewen ?

**Comment by ajlennon** [comment]
Hey @johnmckerrell I found my managed switch. What do you think? Should do the job?

![switch](https://user-images.githubusercontent.com/1537834/57448560-93da8580-7251-11e9-9174-fc7876ec5603.jpg)

**Comment by johnmckerrell** [comment]
Hm.. seems like it might not be fast enough if it’s only got two gigabit ports?

**Comment by ajlennon** [comment]
I know @MatthewCroughan is keen on gigabit but do we really need that for most use-cases?

Perhaps not a long term solution but something that would get us started with minimal expenditure?

**Comment by johnmckerrell** [comment]
2 ports might do to ensure DoES maintains +100 but yes it would mean everyone else would be limited.

**Comment by ajlennon** [comment]
This would just be for the static IPs wouldn't it? I'd be happy enough with this setup for my needs for now.

**Comment by johnmckerrell** [comment]
It would, yes you could bring that in to get us going

**Comment by ajlennon** [comment]
ok boss!

**Comment by johnmckerrell** [comment]
Documentation and stuff for the switch:

https://www.netgear.com/support/product/FS726TP.aspx#Firmware%20Version%202.0.1_14

I've put it in the network cupboard for now. Worth noting that it has PoE ports, we couldn't really use them while the switch is sitting between the Baltic router and our gateway but if we replaced it in the future it could be handy.

**Comment by MatthewCroughan** [comment]
Me and Alex both now have IP addresses. Though we should have a separate secure network for IoT devices and for other people to play with, since as it stands almost anything can and does happen on the flat DoES network. Maybe we can use the remaining IP to do this with? I will make a separate issue.

**Comment by MatthewCroughan** [comment]
This was closed selfishly in the past as I thought *me* personally being able to host services was good enough.

But in fact, it has come to me now via this [twitter thread](https://twitter.com/goatchurch/status/1280080522195197952) that we probably want to host things in the space such as [this lovely dashboard that revspace maintain.](https://revspace.nl/grafiekjes/d/000000001/home?orgId=1&refresh=1m&from=now-90d&to=now) without tunnelling it out of the space or running the whole operation remotely.

I know that this would be easy to solve by having a load balancer in the cloud, or by renting a VPS and hosting an MQTT broker there and having all the sensors connect to it. But there is something about #18 that resonates with me. I think it'd be silly to host those services in the cloud considering the sensors are local to DoES, and when they're down so is everything else anyway, so it's just an unnecessary added cost of $3/month to my bill that doesn't need to exist that provides no benefit other than me not having to ask or advocate for this network configuration.

I recall we have one spare public IP that is not being used by anyone at the moment? Perhaps that could be used?

If we can, I can provide a router running OpenWRT or PFSense which would have a Username/Pass per VLAN to allow people to configure that section of their network. There would be a load balancer [(Traefik)](https://traefik.io) running on this router on port 80/443 that would be configurable, allowing people to reverse proxy like so:

Send your custom domain to the device running at 10.0.100.1 based on the `Host:` header
`example.domain.com -> 10.0.100.1`

This could then be managed as a standard config file hosted here on Github, which people could submit PRs to, which is how people could interact with the machine without SSH access.
